File: src/main/java/de/cplaiz/activecraftcore/modules/ModuleManager.java
package de.cplaiz.activecraftcore.modules;

import de.cplaiz.activecraftcore.ActiveCraftCore;
import de.cplaiz.activecraftcore.exceptions.ModuleException;
import de.cplaiz.activecraftcore.exceptions.OperationFailureException;
import de.cplaiz.activecraftcore.messages.Errors;
import de.cplaiz.activecraftcore.utils.WebReader;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.plugin.InvalidDescriptionException;
import org.bukkit.plugin.InvalidPluginException;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.PluginDescriptionFile;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.*;

public class ModuleManager {

    @Getter
    private static final Set<Module> modules = new HashSet<>();
    private static final String ACPREFIX = "ActiveCraft-";

    static {
        updateModules();
        Bukkit.getScheduler().runTaskTimer(ActiveCraftCore.getInstance(), ModuleManager::updateModules, 20 * 60 * 60, 20 * 60 * 60);
    }

    public static void updateModules() {
        modules.clear();
        try {
            HashMap<String, Integer> acVersionMap = WebReader.getACVersionMap();
            for (String moduleName : acVersionMap.keySet()) {
                if (moduleName.equals(ACPREFIX + "Core")) continue;
                int pluginId = acVersionMap.get(moduleName);
                Map<?, ?> data = WebReader.readAsMap(new URL("https://api.spiget.org/v2/resources/" + pluginId));
                modules.add(new Module(
                        ((String) data.get("name")).replace("ActiveCraft-", ""),
                        pluginId,
                        ((String) data.get("tag")),
                        new URL("https://spigotmc.org/" + ((String) ((Map) data.get("file")).get("url")).split("download\\?")[0]),
                        ((Map) data.get("file")).get("externalUrl") != null ? new URL((String) ((Map) data.get("file")).get("externalUrl")) : new URL("https://spigotmc.org/" + data.get("url"))
                ));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void enable(String moduleName) throws ModuleException {
        Plugin plugin = getLoadedPlugin(moduleName);
        Bukkit.getPluginManager().enablePlugin(plugin);
    }

    public static void load(String moduleName) throws ModuleException {
        checkNotLoaded(moduleName);
        File pluginsDir = new File("plugins");
        if (!pluginsDir.isDirectory())
            pluginsDir.mkdir();
        File pluginFile = null;
        for (File f : pluginsDir.listFiles()) {
            if (!f.getName().endsWith(".jar")) continue;
            PluginDescriptionFile desc;
            try {
                desc = ActiveCraftCore.getInstance().getPluginLoader().getPluginDescription(f);
            } catch (InvalidDescriptionException ignored) {
                throw new ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED);
            }
            if (!desc.getName().equalsIgnoreCase(ACPREFIX + moduleName)) continue;
            pluginFile = f;
            break;
        }
        try {
            Bukkit.getPluginManager().loadPlugin(pluginFile);
        } catch (InvalidPluginException | InvalidDescriptionException e) {
            throw new ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED);
        }
    }

    public static void install(String moduleName) throws OperationFailureException, ModuleException {
        File f = null;
        try {
            f = getInstalled(moduleName);
        } catch (ModuleException ignored) {
        }
        if (f != null) {
            throw new ModuleException(moduleName, ModuleException.ErrorType.ALREADY_INSTALLED);
        }
        try {
            Module module = getModule(moduleName);
            WebReader.downloadFile(module.downloadUrl(),"plugins" + File.separator + ACPREFIX + module.name() + ".jar");
        } catch (IOException e) {
            e.printStackTrace();
            throw new OperationFailureException(Errors.OPERATION_FAILED());
        }
    }

    public static void disable(String moduleName) throws ModuleException {
        Bukkit.getPluginManager().disablePlugin(getEnabledPlugin(moduleName));
    }

    public static Plugin getLoadedPlugin(String moduleName) throws ModuleException {
        getModule(moduleName);
        Plugin plugin = Bukkit.getPluginManager().getPlugin(ACPREFIX + moduleName);
        if (plugin == null)
            throw new ModuleException(moduleName, ModuleException.ErrorType.NOT_LOADED);
        return plugin;
    }

    public static void checkNotLoaded(String moduleName) throws ModuleException {
        getModule(moduleName);
        if (Bukkit.getPluginManager().getPlugin(ACPREFIX + moduleName) != null)
            throw new ModuleException(moduleName, ModuleException.ErrorType.ALREADY_LOADED);
    }

    public static Plugin getEnabledPlugin(String moduleName) throws ModuleException {
        Plugin plugin = getLoadedPlugin(moduleName);
        if (!plugin.isEnabled())
            throw new ModuleException(moduleName, ModuleException.ErrorType.NOT_ENABLED);
        return plugin;
    }

    public static Plugin getDisabledPlugin(String moduleName) throws ModuleException {
        Plugin plugin = getLoadedPlugin(moduleName);
        if (plugin.isEnabled())
            throw new ModuleException(moduleName, ModuleException.ErrorType.ALREADY_ENABLED);
        return plugin;
    }

    public static File getInstalled(String moduleName) throws ModuleException {
        File pluginsDir = new File("plugins");
        if (!pluginsDir.isDirectory())
            pluginsDir.mkdir();
        File pluginFile = null;
        for (File f : pluginsDir.listFiles()) {
            if (!f.getName().endsWith(".jar")) continue;
            PluginDescriptionFile desc;
            try {
                desc = ActiveCraftCore.getInstance().getPluginLoader().getPluginDescription(f);
            } catch (InvalidDescriptionException ignored) {
                throw new ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED);
            }
            if (!desc.getName().equalsIgnoreCase(ACPREFIX + moduleName)) continue;
            return f;
        }
        throw new ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED);
    }

    public static Module getModule(String moduleName) throws ModuleException {
        return modules.stream()
                .filter(module -> module.name().equals(moduleName))
                .findFirst().orElseThrow(() -> new ModuleException(moduleName, ModuleException.ErrorType.DOES_NOT_EXIST));
    }

}

File: src/main/java/org/activecraft/activecraftcore/modules/ModuleManager.kt
package org.activecraft.activecraftcore.modules

import org.activecraft.activecraftcore.ActiveCraftCore
import org.activecraft.activecraftcore.exceptions.ModuleException
import org.activecraft.activecraftcore.exceptions.OperationFailureException
import org.activecraft.activecraftcore.utils.WebReader.aCVersionMap
import org.activecraft.activecraftcore.utils.WebReader.downloadFile
import org.activecraft.activecraftcore.utils.WebReader.readAsMap
import org.bukkit.Bukkit
import org.bukkit.plugin.InvalidDescriptionException
import org.bukkit.plugin.InvalidPluginException
import org.bukkit.plugin.Plugin
import org.bukkit.plugin.PluginDescriptionFile
import java.io.File
import java.io.IOException
import java.net.URL

object ModuleManager {
    val modules: MutableSet<Module> = HashSet()
    private const val ACPREFIX = "ActiveCraft-"

    init {
        updateModules()
        Bukkit.getScheduler().runTaskTimer(
            ActiveCraftCore.INSTANCE,
            Runnable { updateModules() },
            (20 * 60 * 60).toLong(),
            (20 * 60 * 60).toLong()
        )
    }

    fun updateModules() {
        modules.clear()
        try {
            val acVersionMap: Map<String, Int> = aCVersionMap
            for (moduleName in acVersionMap.keys) {
                if (moduleName == ACPREFIX + "Core") continue
                val pluginId = acVersionMap[moduleName]!!
                val data = readAsMap(URL("https://api.spiget.org/v2/resources/$pluginId"))
                modules.add(Module(
                    (data["name"] as String?)!!.replace("ActiveCraft-", ""),
                    pluginId,
                    data["tag"] as String?,
                    URL(
                        "https://spigotmc.org/" + ((data["file"] as Map<*, *>?)!!["url"] as String?)!!.split("download\\?".toRegex())
                            .dropLastWhile { it.isEmpty() }
                            .toTypedArray()[0]),
                    if ((data["file"] as Map<*, *>?)!!["externalUrl"] != null) URL((data["file"] as Map<*, *>?)!!["externalUrl"] as String?) else URL(
                        "https://spigotmc.org/" + data["url"]
                    )
                ))
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Throws(ModuleException::class)
    fun enable(moduleName: String) {
        val plugin = getLoadedPlugin(moduleName)
        Bukkit.getPluginManager().enablePlugin(plugin)
    }

    @Throws(ModuleException::class)
    fun load(moduleName: String) {
        checkNotLoaded(moduleName)
        val pluginsDir = File("plugins")
        if (!pluginsDir.isDirectory) pluginsDir.mkdir()
        var pluginFile: File? = null
        for (f in pluginsDir.listFiles()!!) {
            if (!f.name.endsWith(".jar")) continue
            val desc: PluginDescriptionFile = try {
                ActiveCraftCore.INSTANCE.pluginLoader.getPluginDescription(f)
            } catch (ignored: InvalidDescriptionException) {
                throw ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED)
            }
            if (!desc.name.equals(ACPREFIX + moduleName, ignoreCase = true)) continue
            pluginFile = f
            break
        }
        try {
            Bukkit.getPluginManager().loadPlugin(pluginFile!!)
        } catch (e: InvalidPluginException) {
            throw ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED)
        } catch (e: InvalidDescriptionException) {
            throw ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED)
        }
    }

    @Throws(OperationFailureException::class, ModuleException::class)
    fun install(moduleName: String) {
        var f: File? = null
        try {
            f = getInstalled(moduleName)
        } catch (ignored: ModuleException) {
        }
        if (f != null) {
            throw ModuleException(moduleName, ModuleException.ErrorType.ALREADY_INSTALLED)
        }
        try {
            val module = getModule(moduleName)
            downloadFile(module.downloadUrl, "plugins" + File.separator + ACPREFIX + module.name + ".jar")
        } catch (e: IOException) {
            e.printStackTrace()
            throw OperationFailureException()
        }
    }

    @Throws(ModuleException::class)
    fun disable(moduleName: String) {
        Bukkit.getPluginManager().disablePlugin(getEnabledPlugin(moduleName))
    }

    @Throws(ModuleException::class)
    fun getLoadedPlugin(moduleName: String): Plugin {
        getModule(moduleName)
        return Bukkit.getPluginManager().getPlugin(ACPREFIX + moduleName)
            ?: throw ModuleException(moduleName, ModuleException.ErrorType.NOT_LOADED)
    }

    @Throws(ModuleException::class)
    fun checkNotLoaded(moduleName: String) {
        getModule(moduleName)
        if (Bukkit.getPluginManager().getPlugin(ACPREFIX + moduleName) != null) throw ModuleException(
            moduleName,
            ModuleException.ErrorType.ALREADY_LOADED
        )
    }

    @Throws(ModuleException::class)
    fun getEnabledPlugin(moduleName: String): Plugin {
        val plugin = getLoadedPlugin(moduleName)
        if (!plugin.isEnabled) throw ModuleException(moduleName, ModuleException.ErrorType.NOT_ENABLED)
        return plugin
    }

    @Throws(ModuleException::class)
    fun getDisabledPlugin(moduleName: String): Plugin {
        val plugin = getLoadedPlugin(moduleName)
        if (plugin.isEnabled) throw ModuleException(moduleName, ModuleException.ErrorType.ALREADY_ENABLED)
        return plugin
    }

    @Throws(ModuleException::class)
    fun getInstalled(moduleName: String): File {
        val pluginsDir = File("plugins")
        if (!pluginsDir.isDirectory) pluginsDir.mkdir()
        val pluginFile: File? = null
        for (f in pluginsDir.listFiles()!!) {
            if (!f.name.endsWith(".jar")) continue
            val desc: PluginDescriptionFile = try {
                ActiveCraftCore.INSTANCE.pluginLoader.getPluginDescription(f)
            } catch (ignored: InvalidDescriptionException) {
                throw ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED)
            }
            if (!desc.name.equals(ACPREFIX + moduleName, ignoreCase = true)) continue
            return f
        }
        throw ModuleException(moduleName, ModuleException.ErrorType.NOT_INSTALLED)
    }

    @Throws(ModuleException::class)
    fun getModule(moduleName: String): Module {
        return modules.firstOrNull() { module: Module -> module.name == moduleName }
            ?: throw ModuleException(moduleName, ModuleException.ErrorType.DOES_NOT_EXIST)
    }
}